packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -11,19 +11,55 @@ import 'package:flutter/foundation.dart';
 import 'text_editing.dart';
 import 'text_input.dart';
 
+// Examples can assume:
+// int maxLength;
+
+/// Function signature expected for creating custom [TextInputFormatter]
+/// shorthands via [TextInputFormatter.withFunction].
+typedef TextInputFormatFunction = TextEditingValue Function(
+  TextEditingValue oldValue,
+  TextEditingValue newValue,
+);
+
+/// Function signature for creating a custom
+/// [CompositeTextInputFormatter.shouldReformat] implementation.
+typedef ShouldReformatPredicate = bool Function(
+  TextInputFormatter oldFormatter,
+  CompositeTextInputFormatter newFormatter,
+);
+
 /// A [TextInputFormatter] can be optionally injected into an [EditableText]
 /// to provide as-you-type validation and formatting of the text being edited.
 ///
-/// Text modification should only be applied when text is being committed by the
-/// IME and not on text under composition (i.e., only when
+/// An [EditableText] formats its [TextEditingValue] when the user changes the
+/// text, or when its [EditableText.inputFormatters] parameter changes.
+/// [EditableText] may repetitively apply the same formatter against the input
+/// text, therefore a formatter should not further modify a [TextEditingValue]
+/// if the value has already been formatted by the same formatter.
```

Comment:
But the doc here is actually not entirely correct because even if `f` and `g` are both not idempotent, `g ∘ f` can still be idempotent (e.g., the US phone number example. `EditableText` takes a list of formatters).